File: packages/netsu/src/transport/tcp.ts
import { Socket, connect } from "node:net";
import { ByteBuffer, type BytePipe } from "../protocol/pipe.ts";
import type { DataChannel } from "../streams/channel.ts";

/** Control-channel view of a TCP socket. */
export class TcpPipe implements BytePipe {
  readonly socket: Socket;
  #buffer = new ByteBuffer();
  #onData = (d: Buffer) => this.#buffer.feed(d);
  #onClose = () => this.#buffer.end();
  #detached = false;

  constructor(socket: Socket) {
    this.socket = socket;
    socket.on("data", this.#onData);
    socket.on("close", this.#onClose);
    socket.on("error", this.#onClose);
  }

  readExact(n: number, timeoutMs?: number): Promise<Uint8Array> {
    return this.#buffer.readExact(n, timeoutMs);
  }

  write(data: Uint8Array): Promise<void> {
    if (this.#detached) {
      return Promise.reject(new Error("write on detached TcpPipe"));
    }
    return new Promise((resolve, reject) => {
      this.socket.write(data, (err) => (err ? reject(err) : resolve()));
    });
  }

  /** Stop interpreting bytes; caller takes the raw socket (for TcpDataChannel). */
  detach(): Socket {
    if (this.#buffer.buffered > 0) throw new Error("detach with buffered bytes");
    if (this.#buffer.hasPendingRead) throw new Error("detach with pending readExact");
    this.#detached = true;
    this.socket.off("data", this.#onData);
    this.socket.off("close", this.#onClose);
    this.socket.off("error", this.#onClose);
    return this.socket;
  }

  close(): void {
    // Once detached, the socket belongs to whoever called detach() (typically
    // a TcpDataChannel) — closing here must not reach in and destroy it.
    if (this.#detached) return;
    this.socket.destroy();
    this.#buffer.end();
  }
}

/**
 * Fix 3: without an explicit deadline, connect() to a peer that never
 * completes the TCP handshake (a silently dropped SYN, a firewall that
 * black-holes rather than RSTs) hangs this promise forever with no error at
 * all — mirrors ws.ts's wsConnect fix one layer below the HTTP Upgrade.
 * `socket.setTimeout()` is Node's idiomatic connect-and-idle deadline; it is
 * disabled again as soon as `connect` fires so it doesn't linger as an
 * idle-timeout for the rest of the control channel's life — that's a
 * separate concern, already covered by each readExact() call's own
 * timeoutMs (src/protocol/pipe.ts).
 */
const DEFAULT_CONNECT_TIMEOUT_MS = 10_000;

export function tcpConnect(
  host: string,
  port: number,
  timeoutMs = DEFAULT_CONNECT_TIMEOUT_MS,
): Promise<TcpPipe> {
  return new Promise((resolve, reject) => {
    let settled = false;
    const socket = connect({ host, port, noDelay: true });

    const onTimeout = () => {
      if (settled) return;
      settled = true;
      socket.destroy();
      reject(new Error(`connect timeout after ${timeoutMs}ms`));
    };
    const onError = (err: Error) => {
      if (settled) return;
      settled = true;
      reject(err);
    };
    socket.setTimeout(timeoutMs, onTimeout);
    socket.once("connect", () => {
      if (settled) return;
      settled = true;
      socket.setTimeout(0);
      socket.off("timeout", onTimeout);
      socket.off("error", onError);
      resolve(new TcpPipe(socket));
    });
    socket.once("error", onError);
  });
}

/** Bulk payload over a detached socket. write() honors kernel backpressure via drain. */
export class TcpDataChannel implements DataChannel {
  #socket: Socket;
  /**
   * A write's callback can fail asynchronously after write() already
   * resolved on the fast path (socket.write() returned true). We cannot
   * reject an already-settled promise, so the error is latched here and
   * surfaced — and the channel poisoned — on the next call.
   */
  #pendingError: Error | undefined;
  #dataListenerAttached = false;

  constructor(socket: Socket) {
    this.#socket = socket;
    socket.on("error", (err) => {
      if (!this.#pendingError) this.#pendingError = err;
      socket.destroy();
    });
  }

  write(chunk: Uint8Array): Promise<void> {
    if (this.#pendingError) return Promise.reject(this.#pendingError);
    return new Promise((resolve, reject) => {
      let settled = false;
      const ok = this.#socket.write(chunk, (err) => {
        if (!err) return;
        if (!this.#pendingError) this.#pendingError = err;
        if (!settled) {
          settled = true;
          reject(err);
        }
        // else: this write already resolved on the fast path; the error is
        // latched in #pendingError and will surface on the next write().
      });
      if (ok) {
        settled = true;
        resolve();
      } else {
        this.#socket.once("drain", () => {
          if (!settled) {
            settled = true;
            resolve();
          }
        });
      }
    });
  }

  /** Single-call contract: registers the one "data" listener for this channel. */
  onData(cb: (byteLength: number) => void): void {
    if (this.#dataListenerAttached) {
      throw new Error("TcpDataChannel.onData may only be called once");
    }
    this.#dataListenerAttached = true;
    this.#socket.on("data", (d: Buffer) => cb(d.length));
  }

  /**
   * Exposes a write failure latched in #pendingError so it isn't stranded
   * when the last write of a transfer resolves optimistically and its
   * failure (write callback or socket "error") only arrives after that,
   * with no further write() call to surface it on. Callers finalizing a
   * stream (e.g. after the last write, around close()) must check this.
   */
  get error(): Error | undefined {
    return this.#pendingError;
  }

  close(): void {
    // Bare destroy(): any bytes already sitting in the kernel receive queue
    // that Node hasn't yet emitted as a "data" event are discarded, not
    // drained. This is not only a client-side, reverse-mode concern (an
    // earlier version of this comment understated the scope to just that):
    // src/server.ts's ServerSession#run() destroys every stream, including a
    // forward-mode one where the server itself is the receiver, as soon as
    // it observes TEST_END on the control channel — the same no-drain
    // destroy(), on the other side of the same forward-mode transfer.
    //
    // On the client side specifically: the client no longer destroys its
    // reverse-mode receive streams at duration-timer time (src/client.ts's
    // #startRunning leaves them open so the server can observe TEST_END and
    // stop sending first); the only remaining destroy() there happens in
    // #cleanup()'s finally, after the control-channel handshake completes, by
    // which point the server has already stopped writing. So on the client
    // side the truncation this comment used to document in isolation is no
    // longer expected to occur in practice; retained as a note in case a
    // future caller reintroduces an early destroy() on an active receive
    // channel. The server-side forward-mode case above is a separate,
    // still-live instance of the same underlying risk, not a regression to
    // fix here — this comment is only correcting the scope it describes.
    this.#socket.destroy();
  }
}

File: netsu-rs/src/transport/tcp.rs
//! TCP transport: one socket, two lives.
//!
//! [`TcpPipe`] is the control channel's `BytePipe` — pull-based, used for the
//! cookie/state/JSON handshake. Once the handshake completes,
//! [`TcpPipe::into_data_channel`] hands the same socket to [`TcpDataChannel`]
//! for the bulk test payload, which speaks the (different) `DataChannel`
//! trait instead.
//!
//! Compare with `packages/netsu/src/transport/tcp.ts`: that implementation
//! carries four runtime guards, each documenting a review-found bug.
//! `into_data_channel(self)` here *consumes* the pipe, so two of them —
//! "write after detach" and "close after detach" — are unrepresentable: the
//! caller no longer has a `TcpPipe` to call those methods on, so the type
//! system enforces what the TS version had to check at runtime. The third
//! guard (refusing to detach with bytes still buffered) is kept below,
//! because the protocol's guarantee that no bytes are buffered at that point
//! is exactly that — a guarantee, not a proof the compiler can check, and
//! violating it would silently corrupt the data-channel bytestream. The
//! fourth — the TS connect-timeout fix — is `CONNECT_TIMEOUT` /
//! `TcpPipe::connect`'s `tokio::time::timeout` wrap below.

use std::collections::VecDeque;
use std::time::Duration;

use async_trait::async_trait;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;

use crate::error::{NetsuError, Result};
use crate::protocol::pipe::BytePipe;
use crate::streams::channel::DataChannel;

/// Default connect deadline. The TypeScript implementation shipped without
/// one and an unreachable host hung the connect promise forever; flagged in
/// review and fixed there with `socket.setTimeout`. Here `TcpPipe::connect`
/// wraps the whole resolve-and-connect in `tokio::time::timeout`.
pub const CONNECT_TIMEOUT: Duration = Duration::from_secs(10);

/// Control-channel view of a `TcpStream`: buffers inbound bytes so
/// `read_exact` can pull out exactly the amount the caller asks for,
/// regardless of how the underlying reads happened to chunk.
pub struct TcpPipe {
    stream: TcpStream,
    buf: VecDeque<u8>,
}

impl TcpPipe {
    /// Wraps an already-connected stream, e.g. from `TcpListener::accept`.
    pub fn from_stream(stream: TcpStream) -> Self {
        // Best-effort: control-channel framing is small and latency-sensitive
        // (single state bytes, short JSON), so Nagle's algorithm would add
        // needless delay. Failure here isn't actionable and isn't fatal.
        let _ = stream.set_nodelay(true);
        TcpPipe {
            stream,
            buf: VecDeque::new(),
        }
    }

    /// Connects to `host:port`, honoring `timeout` for both DNS resolution
    /// and the TCP handshake — an unreachable host errors out rather than
    /// hanging forever.
    pub async fn connect(host: &str, port: u16, timeout: Duration) -> Result<TcpPipe> {
        let stream = tokio::time::timeout(timeout, TcpStream::connect((host, port)))
            .await
            .map_err(|_| NetsuError::Timeout)??;
        stream.set_nodelay(true)?;
        Ok(TcpPipe {
            stream,
            buf: VecDeque::new(),
        })
    }

    /// Hands the socket to a [`TcpDataChannel`] for the bulk payload.
    ///
    /// Errors if bytes are still buffered: the protocol guarantees the
    /// handshake leaves none, and silently discarding them here would drop
    /// the start of the data-channel bytestream instead of surfacing the
    /// violated assumption.
    pub fn into_data_channel(self) -> Result<TcpDataChannel> {
        if !self.buf.is_empty() {
            return Err(NetsuError::Protocol(format!(
                "into_data_channel: {} buffered byte(s) would be lost",
                self.buf.len()
            )));
        }
        Ok(TcpDataChannel::new(self.stream))
    }

    // Inherent forwarders so callers can use `TcpPipe` without importing the
    // `BytePipe` trait — the trait stays the abstraction transports are
    // written against generically (see `protocol::framing`'s `read_json` /
    // `write_json`), while a concrete `TcpPipe` remains usable on its own.
    // Fully-qualified syntax below picks the trait impl explicitly; it is
    // not a recursive call to these very methods.

    pub async fn read_exact(&mut self, n: usize, timeout: Option<Duration>) -> Result<Vec<u8>> {
        <Self as BytePipe>::read_exact(self, n, timeout).await
    }

    pub async fn write_all(&mut self, data: &[u8]) -> Result<()> {
        <Self as BytePipe>::write_all(self, data).await
    }

    pub async fn close(&mut self) {
        <Self as BytePipe>::close(self).await
    }
}

impl BytePipe for TcpPipe {
    async fn read_exact(&mut self, n: usize, timeout: Option<Duration>) -> Result<Vec<u8>> {
        let fill = async {
            while self.buf.len() < n {
                let mut chunk = [0u8; 8192];
                let read = self.stream.read(&mut chunk).await?;
                if read == 0 {
                    return Err(NetsuError::PipeClosed);
                }
                self.buf.extend(chunk[..read].iter().copied());
            }
            Ok(self.buf.drain(..n).collect())
        };
        match timeout {
            Some(d) => tokio::time::timeout(d, fill)
                .await
                .map_err(|_| NetsuError::Timeout)?,
            None => fill.await,
        }
    }

    async fn write_all(&mut self, data: &[u8]) -> Result<()> {
        // tokio's AsyncWriteExt::write_all does not return until the kernel
        // has accepted every byte, so this call *is* the backpressure point
        // — no drain-equivalent needed, unlike the Socket.write() callback
        // dance the TS implementation needed (and got wrong twice; see
        // tcp.ts's TcpDataChannel guard comments).
        self.stream.write_all(data).await.map_err(NetsuError::from)
    }

    async fn close(&mut self) {
        let _ = self.stream.shutdown().await;
    }
}

/// Bulk payload channel over a detached `TcpStream`.
pub struct TcpDataChannel {
    stream: TcpStream,
    /// Failure latched so `error()` still reports it during result
    /// finalization even if the caller's last observed call succeeded, or if
    /// they never called this channel again after the failing one. Rust's
    /// `Result` already surfaces a failure to whichever call triggered it —
    /// this exists for the case where that specific `Result` isn't the thing
    /// consulted at teardown time (e.g. a read loop that only branches on
    /// `Ok(0) | Err(_)`, discarding the error itself).
    error: Option<NetsuError>,
}

impl TcpDataChannel {
    fn new(stream: TcpStream) -> Self {
        TcpDataChannel {
            stream,
            error: None,
        }
    }

    /// Records `err`'s message without needing `NetsuError: Clone` (it isn't
    /// — it wraps `std::io::Error`, which isn't `Clone` either). Keeps the
    /// first failure: once poisoned, later errors are almost always the same
    /// underlying cause (writes/reads on an already-broken socket).
    fn latch(&mut self, err: &NetsuError) {
        self.error
            .get_or_insert_with(|| NetsuError::Protocol(err.to_string()));
    }

    // Inherent forwarders, for the same reason `TcpPipe` has them: a caller
    // holding a concrete `TcpDataChannel` shouldn't need to import
    // `DataChannel` just to call its methods by name.

    pub async fn write_chunk(&mut self, chunk: &[u8]) -> Result<()> {
        <Self as DataChannel>::write_chunk(self, chunk).await
    }

    pub async fn read_chunk(&mut self, buf: &mut [u8]) -> Result<usize> {
        <Self as DataChannel>::read_chunk(self, buf).await
    }

    pub async fn close(&mut self) {
        <Self as DataChannel>::close(self).await
    }
}

#[async_trait]
impl DataChannel for TcpDataChannel {
    async fn write_chunk(&mut self, chunk: &[u8]) -> Result<()> {
        if let Some(err) = &self.error {
            return Err(NetsuError::Protocol(err.to_string()));
        }
        match self.stream.write_all(chunk).await {
            Ok(()) => Ok(()),
            Err(e) => {
                let err = NetsuError::from(e);
                self.latch(&err);
                Err(err)
            }
        }
    }

    async fn read_chunk(&mut self, buf: &mut [u8]) -> Result<usize> {
        if let Some(err) = &self.error {
            return Err(NetsuError::Protocol(err.to_string()));
        }
        match self.stream.read(buf).await {
            Ok(n) => Ok(n),
            Err(e) => {
                let err = NetsuError::from(e);
                self.latch(&err);
                Err(err)
            }
        }
    }

    async fn close(&mut self) {
        // Bare shutdown: any bytes already in the kernel receive queue that
        // we haven't read yet are simply left unread, not an error — the
        // control channel has already told both sides the test is over by
        // the time close() is called.
        let _ = self.stream.shutdown().await;
    }

    fn error(&self) -> Option<&NetsuError> {
        self.error.as_ref()
    }
}
